x-pack/plugins/security_solution/public/management/services/exceptions_list/exceptions_list_api_client.ts

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  CreateExceptionListItemSchema,
+  CreateExceptionListSchema,
+  ExceptionListItemSchema,
+  ExceptionListSummarySchema,
+  FoundExceptionListItemSchema,
+  UpdateExceptionListItemSchema,
+} from '@kbn/securitysolution-io-ts-list-types';
+import { EXCEPTION_LIST_ITEM_URL, EXCEPTION_LIST_URL } from '@kbn/securitysolution-list-constants';
+import { HttpStart } from 'kibana/public';
+
+/**
+ * A generic class to be used for each artifact type.
+ * It follow the Singleton pattern.
+ * Please, use the getInstance method instead of creating a new instance when using this implementation.
+ */
+export class ExceptionsListApiClient {
+  private static instance: ExceptionsListApiClient;
+  private ensureListExists: Promise<void>;
+
+  constructor(
+    private readonly http: HttpStart,
+    private readonly listId: string,
+    private readonly listDefinition: CreateExceptionListSchema
+  ) {
+    this.ensureListExists = this.createExceptionList();
+  }
+
+  /**
+   * PrivateStatic method that creates the list and don't throw if list already exists.
+   * This method is being used when initializing an instance only once.
+   */
+  private async createExceptionList(): Promise<void> {
+    try {
+      await this.http.post<ExceptionListItemSchema>(EXCEPTION_LIST_URL, {
```

Comment:
This here is where I think we can introduce local module state to keep track if we already created the list. so above (lets say - on line 18; outside of the class), create:

```
const wasListCreated: Record<string, Promise<void>> = {};
```

Then this method could look like this:

```typescript
async createExceptionList(): Promise<void> {
    const listId = this.listId;

    // if list is already created or in the process of being created, then return that promise
    if (wasListCreated[listId]) {
      return wasListCreated[listId]
    }
    
    // Store a Promise that tracks if the list is created or being created. This will ensure that
    // if this method (`craeteExceptionList()`) is called multiple times for the same list ID, we
    // only dispatch one HTTP request.
    wasListCreated[listId] = new Promise<void>((resolve, reject) => {
      try {
        await this.http.post<ExceptionListItemSchema>(
          EXCEPTION_LIST_URL,
          {body: JSON.stringify({ ...this.listDefinition, list_id: this.listId }),
        });

        resolve();
      } catch (err) {
        // Ignore 409 errors. List already created
        if (err.response?.status !== 409) {
          reject(err);
        }
        
        resolve();
      }
    });

    // If creation fails, then remove the list id from the `wasListCreated` object.
    wasListCreated[listId].catch(() => {
      delete wasListCreated[listId];
    });

    return wasListCreated[listId];
  }

```

If you decide to go with this approach, then I would say that `this.ensureListExists` should be removed and instead, this function should be renamed `ensureListExists()`, because you know it will only ever attempt to create the list once, unless there was a failure